packages/js/src/executors/tsc/lib/batch/typescript-compilation.ts

```diff
@@ -0,0 +1,305 @@
+import type { TaskGraph } from '@nx/devkit';
+import { logger } from '@nx/devkit';
+import { createAsyncIterable } from '@nx/devkit/src/utils/async-iterable';
+import type { BatchResults } from 'nx/src/tasks-runner/batch/batch-messages';
+import * as ts from 'typescript';
+import { getCustomTrasformersFactory } from '../get-custom-transformers-factory';
+import type { TaskInfo } from './types';
+import {
+  formatDiagnosticReport,
+  formatSolutionBuilderStatusReport,
+  formatWatchStatusReport,
+} from './typescript-diagnostic-reporters';
+
+// https://github.com/microsoft/TypeScript/blob/d45012c5e2ab122919ee4777a7887307c5f4a1e0/src/compiler/diagnosticMessages.json#L4050-L4053
+// Typescript diagnostic message for 5083: Cannot read file '{0}'.
+const TYPESCRIPT_CANNOT_READ_FILE = 5083;
+// https://github.com/microsoft/TypeScript/blob/d45012c5e2ab122919ee4777a7887307c5f4a1e0/src/compiler/diagnosticMessages.json#L4211-4214
+// Typescript diagnostic message for 6032: File change detected. Starting incremental compilation...
+const TYPESCRIPT_FILE_CHANGE_DETECTED_STARTING_INCREMENTAL_COMPILATION = 6032;
+
+const now = () => new Date().getTime();
+
+export function compileBatchTypescript(
+  tsConfigTaskInfoMap: Record<string, TaskInfo>,
+  taskGraph: TaskGraph,
+  watch: boolean,
+  postProjectCompilationCallback: (taskInfo: TaskInfo) => void
+): {
+  iterator: AsyncIterable<BatchResults>;
+  close: () => void | Promise<void>;
+} {
+  const timeNow = now();
+  const defaultResults: BatchResults = Object.keys(taskGraph.tasks).reduce(
+    (acc, task) => {
+      acc[task] = { success: true, startTime: timeNow, terminalOutput: '' };
+      return acc;
+    },
+    {} as BatchResults
+  );
+
+  let tearDown: (() => void) | undefined;
+
+  return {
+    iterator: createAsyncIterable<BatchResults>(({ next, done }) => {
+      if (watch) {
+        compileTSWithWatch(tsConfigTaskInfoMap, postProjectCompilationCallback);
+
+        tearDown = () => {
+          // watch runs are not cached and batch executors are not composable,
+          // results here are not meaningful, just return something
+          next(defaultResults);
```

Comment:
Since in watch mode, the `process.exit(...)` is called on SIGINT and SIGTERM, the result here are never used. Is that right?

Reply:
This is the same thing done in the non-batch executor and I just replicated it, but I agree it's not needed. I'll remove it.